mlrun/runtimes/pod.py

```diff
@@ -411,6 +431,316 @@ def enrich_resources_with_default_pod_resources(
+                        > Adds anti-affinity IF no tolerations were configured
+        """
+        # nothing to do here, configuration is not populated
+        if (
+            not mlconf.get_preemptible_tolerations()
+            and not mlconf.get_preemptible_node_selector()
+        ):
+            return
+
+        if not self.preemption_mode:
+            # We're not supposed to get here, but if we do, we'll set the private attribute to
+            # avoid triggering circular enrichment.
+            self._preemption_mode = mlconf.function_defaults.preemption_mode
+            logger.debug(
+                "No preemption mode was given, using the default preemption mode",
+                default_preemption_mode=self.preemption_mode,
+            )
+
+        logger.debug(
+            "Enriching function spec for given preemption mode",
+            preemption_mode=self.preemption_mode,
+        )
+        # remove preemptible tolerations and remove preemption related configuration
+        # and enrich with anti-affinity if preemptible tolerations configuration haven't been provided
+        if self.preemption_mode == PreemptionModes.prevent:
+            # ensure no preemptible node tolerations
+            self._prune_tolerations(generate_preemptible_tolerations())
+            self._clear_tolerations_if_initialized_but_empty()
+
+            # if preemptible tolerations were given, purge affinity preemption related configuration
+            self._prune_affinity_node_selector_requirement(
+                (
+                    generate_preemptible_node_selector_requirements(
+                        NodeSelectorOperator.node_selector_op_in
+                    )
+                )
+            )
+            # using a single term with potentially multiple expressions to ensure affinity
+            self._merge_node_selector_term_to_node_affinity(
+                node_selector_terms=generate_preemptible_nodes_anti_affinity_terms()
+            )
+
+        # enrich tolerations and override all node selector terms with preemptible node selector terms
+        elif self.preemption_mode == PreemptionModes.constrain:
+            # enrich with tolerations
+            self._merge_tolerations(generate_preemptible_tolerations())
+            self._clear_tolerations_if_initialized_but_empty()
+
+            self._initialize_affinity()
+            self._initialize_node_affinity()
+            # setting required_during_scheduling_ignored_during_execution
+            # overriding other terms that have been set, and only setting terms for preemptible nodes
+            # when having multiple terms, pod scheduling is succeeded if at least one term is satisfied
+            self.affinity.node_affinity.required_during_scheduling_ignored_during_execution = k8s_client.V1NodeSelector(
+                node_selector_terms=generate_preemptible_nodes_affinity_terms()
+            )
```

Comment:
We don't want to merge we want to override, if we merge than it could be scheduled on not preemptible nodes, in this block we want to run **only** on preemptible nodes.